src/gui/base/WizardDialog.ts

```diff
@@ -264,42 +306,63 @@ export type WizardDialogAttrsBuilder<T> = {
 	attrs: WizardDialogAttrs<T>
 }
 
+const headerBarAttrs: DialogHeaderBarAttrs = {}
+
 // Use to generate a new wizard
 export function createWizardDialog<T>(data: T, pages: ReadonlyArray<WizardPageWrapper<T>>, closeAction?: () => $Promisable<void>): WizardDialogAttrsBuilder<T> {
 	// We need the close action of the dialog before we can create the proper attributes
-	const headerBarAttrs: DialogHeaderBarAttrs = {}
 
 	let view: () => Children = () => null
 
 	const child: Component = {
 		view: () => view(),
 	}
+	const unregisterCloseListener = windowFacade.addWindowCloseListener(() => {})
 	const wizardDialog = Dialog.largeDialog(headerBarAttrs, child)
-	const wizardDialogAttrs = new WizardDialogAttrs(
-		data,
-		pages,
-		closeAction ? () => Promise.resolve(closeAction()).then(() => wizardDialog.close()) : async () => wizardDialog.close(),
-	)
-	// We replace the dummy values from dialog creation
-	const wizardDialogHeaderBarAttrs: DialogHeaderBarAttrs = wizardDialogAttrs.getHeaderBarAttrs()
-	Object.entries(wizardDialogHeaderBarAttrs).forEach(([key, value]) => {
-		// @ts-ignore
-		headerBarAttrs[key] = value
-	})
+	const closeActionWrapper = async () => {
+		if (closeAction) {
+			await closeAction()
+		}
+		wizardDialog.close()
+		unregisterCloseListener()
+	}
+	const wizardDialogAttrs = new WizardDialogAttrs(data, pages, closeActionWrapper)
+	renderHeader(wizardDialogAttrs.getHeaderBarAttrs())
 
 	view = () => m(WizardDialog, wizardDialogAttrs)
-
 	wizardDialog
 		.addShortcut({
 			key: Keys.ESC,
 			exec: () => {
-				wizardDialogAttrs.closeAction()
+				confirmThenCleanup(() => wizardDialogAttrs.closeAction())
 			},
 			help: "close_alt",
 		})
-		.setCloseHandler(() => wizardDialogAttrs.goToPreviousPageOrClose())
+		.setCloseHandler(() => {
+			// the dialogs popState handler will return false which prevents the wizard from being closed
+			// we then close the wizard manually if the user confirms
+			confirmThenCleanup(() => wizardDialogAttrs.closeAction())
+		})
+
 	return {
 		dialog: wizardDialog,
 		attrs: wizardDialogAttrs,
 	}
+
+	async function confirmThenCleanup(closeAction: () => Promise<void>) {
```

Comment:
this function can be moved to the top level

Reply:
That's true, I also decided to not unregister the close listener in this method since this is also always done at the end of the closeAction(). Do you have any objections? Currently it is kinda duplicated.